File: assignment 8/c/programs/2.c
#include <stdio.h>

// Function to perform binary search
int binarySearch(int arr[], int size, int target)
{
    int left = 0;
    int right = size - 1;

    while (left <= right)
    {
        int mid = left + (right - left) / 2;

        if (arr[mid] == target)
        {
            return mid; // Return the index where the target is found
        }
        else if (arr[mid] < target)
        {
            left = mid + 1;
        }
        else
        {
            right = mid - 1;
        }
    }

    return -1; // Return -1 if the target is not found
}

int main()
{
    int arr[] = {2, 5, 7, 10, 12, 15, 20};
    int size = sizeof(arr) / sizeof(arr[0]);
    int target = 12;

    int result = binarySearch(arr, size, target);

    if (result != -1)
    {
        printf("Element %d found at index %d.\n", target, result);
    }
    else
    {
        printf("Element %d not found in the array.\n", target);
    }

    return 0;
}

File: assignment 8/python/programs/2.py
def binary_search(arr, target):
    left = 0
    right = len(arr) - 1

    while left <= right:
        mid = (left + right) // 2  # Calculate the middle index

        if arr[mid] == target:
            return mid  # Target found, return the index
        elif arr[mid] < target:
            left = mid + 1  # Target is in the right half of the remaining array
        else:
            right = mid - 1  # Target is in the left half of the remaining array

    return -1  # Target not found in the list


# Example usage:
my_list = [3, 6, 8, 12, 15, 24, 35, 48]
target_element = 15

# Binary search requires the list to be sorted
my_list.sort()

result = binary_search(my_list, target_element)

if result != -1:
    print(f"Element {target_element} found at index {result}")
else:
    print(f"Element {target_element} not found in the list.")
